File: src/engine/session.py
"""Which unit next, and which of its cards: the scheduler over the deck.

Due units first, least likely to be remembered first; then new units in
rank order, skipping trivial and known ones, up to ``new_per_day``; then
units due within the learn-ahead window. A card is shown only when it has
a machine gloss (rule 9) and, for a sentence-initial connector, only when it
has its preceding sentence (a bare "Trotzdem kam sie." is a guess, not an
exercise) unless the unit has no other card.
"""

from __future__ import annotations

from collections.abc import Callable, Sequence
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from pathlib import Path

from src.contracts import DeckManifest, PhraseCard, PhraseUnit
from src.engine.fsrs import FSRSEngine
from src.engine.review_log import LearnerState
from src.phrases.export import DEFAULT_DECK_DIR, load_deck


@dataclass(frozen=True)
class Settings:
    """``cards_per_day`` is the day's exercise budget: due units always run,
    new units are introduced until that many answers have been given today.
    ``new_per_day`` is a hard cap on top of it, ``None`` for none. The first
    learner found a cap of ten units repetitive (feedback, 2026-09-11)."""

    cards_per_day: int = 40
    new_per_day: int | None = None
    learn_ahead: timedelta = timedelta(minutes=20)
    retention: float = 0.9


@dataclass
class Deck:
    manifest: DeckManifest
    units: list[PhraseUnit]
    cards_by_unit: dict[str, list[PhraseCard]]

    @property
    def by_id(self) -> dict[str, PhraseUnit]:
        return {u.unit_id: u for u in self.units}

    @classmethod
    def load(cls, deck_dir: Path = DEFAULT_DECK_DIR) -> Deck:
        manifest, units, cards = load_deck(deck_dir)
        by_unit: dict[str, list[PhraseCard]] = {}
        for card in cards:
            by_unit.setdefault(card.unit_id, []).append(card)
        return cls(
            manifest=manifest, units=sorted(units, key=lambda u: u.rank), cards_by_unit=by_unit
        )


def showable_cards(cards: Sequence[PhraseCard]) -> list[PhraseCard]:
    glossed = [c for c in cards if c.gloss_en is not None]
    with_context = [c for c in glossed if not c.needs_context or c.context_de is not None]
    return with_context or glossed


def is_learnable(deck: Deck, unit: PhraseUnit, state: LearnerState) -> bool:
    return (
        not unit.trivial
        and unit.unit_id not in state.known
        and bool(showable_cards(deck.cards_by_unit.get(unit.unit_id, [])))
    )


def new_units_started_today(state: LearnerState, now: datetime) -> int:
    day = now.astimezone(UTC).date()
    return sum(1 for ts in state.first_review.values() if ts.astimezone(UTC).date() == day)


def reviews_today(state: LearnerState, now: datetime) -> int:
    day = now.astimezone(UTC).date()
    return sum(1 for ts in state.review_times if ts.astimezone(UTC).date() == day)


def budget_left(state: LearnerState, settings: Settings, now: datetime) -> int:
    return max(settings.cards_per_day - reviews_today(state, now), 0)


def due_units(
    deck: Deck, state: LearnerState, engine: FSRSEngine, now: datetime, horizon: timedelta
) -> list[PhraseUnit]:
    """Reviewed units whose due time is within ``horizon`` of now, least
    retrievable first."""
    by_id = deck.by_id
    due: list[tuple[float, int, PhraseUnit]] = []
    for unit_id, record in state.records.items():
        unit = by_id.get(unit_id)
        if unit is None or unit_id in state.known or record.due > now + horizon:
            continue
        due.append((engine.get_retrievability(record, now), unit.rank, unit))
    due.sort(key=lambda item: (item[0], item[1]))
    return [unit for _, _, unit in due]


def _not_last(units: list[PhraseUnit], last: str | None) -> list[PhraseUnit]:
    """Never the unit just shown when anything else is available."""
    others = [u for u in units if u.unit_id != last]
    return others or units


def next_unit(
    deck: Deck,
    state: LearnerState,
    engine: FSRSEngine,
    settings: Settings,
    now: datetime,
    *,
    over_limit: bool = False,
) -> PhraseUnit | None:
    """Due units first, least retrievable first; then new units in rank order
    while the day's budget lasts (or ``over_limit`` says go on); then the
    learn-ahead window. The unit shown last is skipped when there is a
    choice, so a session never shows one unit twice in a row."""
    last = state.last_unit
    overdue = _not_last(due_units(deck, state, engine, now, timedelta(0)), last)
    if overdue:
        return overdue[0]
    within_budget = over_limit or budget_left(state, settings, now) > 0
    under_cap = settings.new_per_day is None or (
        new_units_started_today(state, now) < settings.new_per_day
    )
    if within_budget and (under_cap or over_limit):
        for unit in deck.units:
            if unit.unit_id in state.records:
                continue
            if is_learnable(deck, unit, state):
                return unit
    # Learn-ahead never re-shows the unit just answered: that is the
    # back-to-back repeat the first learner complained about.
    ahead = [
        u for u in due_units(deck, state, engine, now, settings.learn_ahead) if u.unit_id != last
    ]
    return ahead[0] if ahead else None


def units_by_stage(
    deck: Deck, state: LearnerState, now: datetime
) -> dict[str, list[tuple[PhraseUnit, datetime | None]]]:
    """Every unit the log knows, grouped: known, deferred, learning, young,
    mature; each with its next due time (``None`` for known and deferred)."""
    by_id = deck.by_id
    groups: dict[str, list[tuple[PhraseUnit, datetime | None]]] = {
        "known": [],
        "deferred": [],
        "learning": [],
        "young": [],
        "mature": [],
    }
    for unit_id in state.known:
        if unit_id in by_id:
            stage = "deferred" if state.known_source.get(unit_id) == "defer" else "known"
            groups[stage].append((by_id[unit_id], None))
    for unit_id, record in state.records.items():
        unit = by_id.get(unit_id)
        if unit is None or unit_id in state.known:
            continue
        if record.state != "review" or record.stability is None:
            groups["learning"].append((unit, record.due))
        elif record.stability < 21.0:
            groups["young"].append((unit, record.due))
        else:
            groups["mature"].append((unit, record.due))
    for items in groups.values():
        items.sort(key=lambda pair: (pair[1] or now, pair[0].rank))
    return groups


def is_praeteritum(card: PhraseCard) -> bool:
    """A finite past-tense form in the gaps (``form_key`` ``Fin|Past|...``)."""
    return card.form_key.startswith("Fin|Past")


def pick_card(
    deck: Deck, unit: PhraseUnit, state: LearnerState, choose: Callable[[int], int] | None = None
) -> PhraseCard | None:
    """Rotate through the unit's showable cards, never the one shown last.
    ``choose(n)`` picks an index in ``range(n)``; the default is the card
    after the last one, so a session walks the surface forms in order."""
    cards = showable_cards(deck.cards_by_unit.get(unit.unit_id, []))
    record = state.records.get(unit.unit_id)
    if record is None or record.state != "review":
        # Präteritum is a B1 form; a unit still being learnt is shown in the
        # present, the perfect and the infinitive first (feedback 2026-09-13).
        easier = [c for c in cards if not is_praeteritum(c)]
        cards = easier or cards
    if not cards:
        return None
    if len(cards) == 1:
        return cards[0]
    last = state.last_card.get(unit.unit_id)
    ids = [c.card_id for c in cards]
    if choose is not None:
        candidates = [c for c in cards if c.card_id != last]
        return candidates[choose(len(candidates)) % len(candidates)]
    index = (ids.index(last) + 1) % len(cards) if last in ids else 0
    return cards[index]


def untriaged_units(deck: Deck, state: LearnerState) -> list[PhraseUnit]:
    """Rank order, skipping trivial units and units already judged or
    already reviewed."""
    return [
        u
        for u in deck.units
        if not u.trivial and u.unit_id not in state.triaged and u.unit_id not in state.records
    ]

File: web/lib/session.js
// The scheduler and stats, matching src/engine/session.py and stats.py.

export const DEFAULT_SETTINGS = { cardsPerDay: 40, newPerDay: null, learnAheadMinutes: 20, retention: 0.9 };

const dayOf = (iso) => new Date(iso).toISOString().slice(0, 10);

export function showableCards(cards) {
  const glossed = cards.filter((c) => c.gloss_en !== null && c.gloss_en !== undefined);
  const withContext = glossed.filter((c) => !c.needs_context || c.context_de);
  return withContext.length ? withContext : glossed;
}

export function isLearnable(deck, unit, state) {
  return !unit.trivial && !state.known.has(unit.unit_id) && showableCards(deck.cardsByUnit[unit.unit_id] || []).length > 0;
}

export function reviewsToday(state, now) {
  const day = dayOf(now);
  return state.reviewTimes.filter((ts) => dayOf(ts) === day).length;
}

export function newUnitsStartedToday(state, now) {
  const day = dayOf(now);
  return Object.values(state.firstReview).filter((ts) => dayOf(ts) === day).length;
}

export function budgetLeft(state, settings, now) {
  return Math.max(settings.cardsPerDay - reviewsToday(state, now), 0);
}

export function dueUnits(deck, state, engine, now, horizonMs) {
  const limit = new Date(now).getTime() + horizonMs;
  const due = [];
  for (const [unitId, record] of Object.entries(state.records)) {
    const unit = deck.byId[unitId];
    if (!unit || state.known.has(unitId) || new Date(record.due).getTime() > limit) continue;
    due.push([engine.retrievability(record, now), unit.rank, unit]);
  }
  due.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  return due.map((d) => d[2]);
}

function notLast(units, last) {
  const others = units.filter((u) => u.unit_id !== last);
  return others.length ? others : units;
}

export function nextUnit(deck, state, engine, settings, now, { overLimit = false } = {}) {
  const last = state.lastUnit;
  const overdue = notLast(dueUnits(deck, state, engine, now, 0), last);
  if (overdue.length) return overdue[0];
  const withinBudget = overLimit || budgetLeft(state, settings, now) > 0;
  const underCap = settings.newPerDay === null || newUnitsStartedToday(state, now) < settings.newPerDay;
  if (withinBudget && (underCap || overLimit)) {
    for (const unit of deck.units) {
      if (unit.unit_id in state.records) continue;
      if (isLearnable(deck, unit, state)) return unit;
    }
  }
  const ahead = dueUnits(deck, state, engine, now, settings.learnAheadMinutes * 60000).filter((u) => u.unit_id !== last);
  return ahead.length ? ahead[0] : null;
}

// A finite past-tense form in the gaps: Präteritum is a B1 form, so a unit
// still being learnt is shown in the present, the perfect and the infinitive
// first (feedback 2026-09-13). Matches src/engine/session.py.
export function isPraeteritum(card) { return (card.form_key || "").startsWith("Fin|Past"); }

export function pickCard(deck, unit, state) {
  let cards = showableCards(deck.cardsByUnit[unit.unit_id] || []);
  const record = state.records[unit.unit_id];
  if (!record || record.state !== "review") {
    const easier = cards.filter((c) => !isPraeteritum(c));
    cards = easier.length ? easier : cards;
  }
  if (!cards.length) return null;
  if (cards.length === 1) return cards[0];
  const last = state.lastCard[unit.unit_id];
  const idx = cards.findIndex((c) => c.card_id === last);
  return cards[idx >= 0 ? (idx + 1) % cards.length : 0];
}

export function untriagedUnits(deck, state) {
  return deck.units.filter((u) => !u.trivial && !state.triaged.has(u.unit_id) && !(u.unit_id in state.records));
}

export function unitsByStage(deck, state, now) {
  const groups = { known: [], deferred: [], learning: [], young: [], mature: [] };
  for (const id of state.known) {
    if (!deck.byId[id]) continue;
    groups[state.knownSource[id] === "defer" ? "deferred" : "known"].push([deck.byId[id], null]);
  }
  for (const [id, r] of Object.entries(state.records)) {
    const unit = deck.byId[id];
    if (!unit || state.known.has(id)) continue;
    if (r.state !== "review" || r.stability === null) groups.learning.push([unit, r.due]);
    else if (r.stability < 21) groups.young.push([unit, r.due]);
    else groups.mature.push([unit, r.due]);
  }
  for (const items of Object.values(groups)) {
    items.sort((a, b) => ((a[1] || now) < (b[1] || now) ? -1 : (a[1] || now) > (b[1] || now) ? 1 : a[0].rank - b[0].rank));
  }
  return groups;
}

export function computeStats(deck, state, entries, engine, now) {
  const s = { known: state.known.size, learning: 0, young: 0, mature: 0, due_now: 0, new_remaining: 0, reviews_today: 0, reviews_total: 0, streak_days: 0, retention_30d: null, coverage: [] };
  const nowMs = new Date(now).getTime();
  for (const [id, r] of Object.entries(state.records)) {
    if (state.known.has(id) || !deck.byId[id]) continue;
    if (r.state !== "review" || r.stability === null) s.learning++;
    else if (r.stability < 21) s.young++;
    else s.mature++;
    if (new Date(r.due).getTime() <= nowMs) s.due_now++;
  }
  s.new_remaining = deck.units.filter((u) => !(u.unit_id in state.records) && isLearnable(deck, u, state)).length;
  const reviews = entries.filter((e) => e.type === "review");
  s.reviews_total = reviews.length;
  const days = new Map();
  for (const e of reviews) days.set(dayOf(e.ts), (days.get(dayOf(e.ts)) || 0) + 1);
  const today = dayOf(now);
  s.reviews_today = days.get(today) || 0;
  let d = new Date(today);
  if (!days.has(today)) d = new Date(d.getTime() - 86400000);
  while (days.has(d.toISOString().slice(0, 10))) { s.streak_days++; d = new Date(d.getTime() - 86400000); }
  const recent = reviews.filter((e) => new Date(e.ts).getTime() >= nowMs - 30 * 86400000);
  if (recent.length) s.retention_30d = Math.round((recent.filter((e) => e.rating !== "again").length / recent.length) * 1000) / 1000;
  const bands = new Map();
  for (const u of deck.units) {
    const band = Math.floor((u.rank - 1) / 500) * 500 + 1;
    const b = bands.get(band) || { band, seen: 0, total: 0 };
    b.total++;
    if (u.unit_id in state.records || state.known.has(u.unit_id)) b.seen++;
    bands.set(band, b);
  }
  s.coverage = [...bands.values()].sort((a, b) => a.band - b.band);
  return s;
}
